tensorflow/compiler/xla/service/gpu/gemm_rewriter.cc

```diff
@@ -241,25 +241,29 @@ class GemmRewriterVisitor : public DfsHloRewriteVisitor {
   }
 
   Status HandleAdd(HloInstruction *instr) override {
-    HloInstruction *bias, *existing_gemm, *existing_slice;
-    // Attempt to elide broadcast and fuse addition of a vector bias into GEMM.
-    if (Match(instr, m::AddAnyOrder(gemm_cublaslt(&existing_gemm),
+    HloInstruction *bias, *existing_gemm, *optional_slice;
+    // Attempt to elide broadcast and fuse addition of a vector bias into GEMM,
+    // including when slicing is applied to the result.
+    if (Match(instr, m::AddAnyOrder(m::OptionalUnaryOp(
+                                        &optional_slice, {HloOpcode::kSlice},
+                                        gemm_cublaslt(&existing_gemm))
+                                        .WithOneUser(),
                                     broadcast_bias(&bias)))) {
-      TF_ASSIGN_OR_RETURN(bool was_fused,
-                          FuseVectorBiasAdd(instr, bias, existing_gemm));
-      if (was_fused) return OkStatus();
-    }
-
-    // Attempt to elide broadcast and fuse addition of a vector bias into GEMM
-    // when slicing is applied to the result.
-    if (Match(instr, m::AddAnyOrder(
-                         slice(&existing_slice, gemm_cublaslt(&existing_gemm)),
-                         broadcast_bias(&bias)))) {
-      TF_ASSIGN_OR_RETURN(
-          bool was_fused,
-          FuseVectorBiasAdd(instr, bias, existing_gemm, existing_slice));
-      if (was_fused) {
-        return OkStatus();
+      switch (optional_slice->opcode()) {
```

Comment:
This is not how OptionalUnaryOp was implemented, see e.g. [this test](https://github.com/tensorflow/tensorflow/blob/3cda567ac1cae720a9fa1cdb0443f4e2e22bfd5d/tensorflow/compiler/xla/service/pattern_matcher_test.cc#L1170-L1173). However, with the switch removed, this doesn't seem to negatively affect the current change.

Reply:
My intent is to build infrastructure that works for this PR, but also works for everyone else who hacks on XLA.  Part of "works" means that it's easy for them to understand, and that it behaves in the way they expect.

I think the current behavior will be surprising to most XLA developers and therefore it should be changed.

Reply:
To confirm, in the test linked above, Match should still return true, but instr should be nullptr? To implement this change, can we move the capture of the instruction pointer to the Match call in HloInstructionPatternOptionalUnaryOpImpl?

Reply:
OK, I looked at how one would implement this.  It seems very complicated.  I don't really see how to move the capture into HloInstructionPatternOptionalUnaryOpImpl, because the capture is actually handled by Op().  :-/

I think you're off the hook on this one, Philipp.